File: polyglot/csharp/config_parser.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Threading.Tasks;

namespace polyglot.csharp
{
    // =====================================================================
    // CONFIGURATION SCHEMA
    // =====================================================================

    /// <summary>
    /// Root configuration object for dastlite.
    /// </summary>
    public sealed class Config
    {
        [JsonPropertyName("targets")]
        public List<UrlTarget> Targets { get; set; } = new();

        [JsonPropertyName("authentication")]
        public AuthConfig Authentication { get; set; } = default!;

        [JsonPropertyName("ruleset")]
        public RulesetConfig Ruleset { get; set; } = default!;

        [JsonPropertyName("output")]
        public OutputConfig Output { get; set; } = default!;

        [JsonPropertyName("limits")]
        public LimitsConfig Limits { get; set; } = default!;

        /// <summary>
        /// Whether to enable verbose logging.
        /// </summary>
        [JsonPropertyName("verbose")]
        public bool Verbose { get; set; } = false;

        /// <summary>
        /// Environment variable substitution prefix (e.g., "env:").
        /// </summary>
        [JsonPropertyName("env_prefix")]
        public string EnvPrefix { get; set; } = "env:";
    }

    /// <summary>
    /// URL target definition.
    /// </summary>
    public sealed class UrlTarget
    {
        [JsonPropertyName("url")]
        public required string Url { get; set; }

        [JsonPropertyName("method")]
        public string Method { get; set; } = "GET";

        [JsonPropertyName("headers")]
        public Dictionary<string, string> Headers { get; set; } = new();

        [JsonPropertyName("params")]
        public Dictionary<string, string> Params { get; set; } = new();

        [JsonPropertyName("body")]
        public string Body { get; set; } = null!;

        /// <summary>
        /// Whether this is a mobile API target.
        /// </summary>
        [JsonPropertyName("mobile_api")]
        public bool MobileApi { get; set; } = false;
    }

    /// <summary>
    /// Authentication configuration.
    /// </summary>
    public sealed class AuthConfig
    {
        [JsonPropertyName("type")]
        public required string Type { get; set; }

        [JsonPropertyName("api_key")]
        public string ApiKey { get; set; } = null!;

        [JsonPropertyName("bearer_token")]
        public string BearerToken { get; set; } = null!;

        [JsonPropertyName("oauth_client_id")]
        public string OAuthClientId { get; set; } = null!;

        [JsonPropertyName("oauth_client_secret")]
        public string OAuthClientSecret { get; set; } = null!;

        /// <summary>
        /// Whether to use session cookies for auth.
        /// </summary>
        [JsonPropertyName("session_cookie")]
        public bool SessionCookie { get; set; } = false;
    }

    /// <summary>
    /// Ruleset configuration - defines what scans to run.
    /// </summary>
    public sealed class RulesetConfig
    {
        [JsonPropertyName("enabled")]
        public bool Enabled { get; set; } = true;

        [JsonPropertyName("rules")]
        public List<RuleDefinition> Rules { get; set; } = new();

        /// <summary>
        /// Whether to run in "discover" mode (auto-detect endpoints).
        /// </summary>
        [JsonPropertyName("discover_mode")]
        public bool DiscoverMode { get; set; } = false;

        /// <summary>
        /// Custom headers to inject into all requests.
        /// </summary>
        [JsonPropertyName("global_headers")]
        public Dictionary<string, string> GlobalHeaders { get; set; } = new();
    }

    /// <summary>
    /// Individual rule definition.
    /// </summary>
    public sealed class RuleDefinition
    {
        [JsonPropertyName("id")]
        public required string Id { get; set; }

        [JsonPropertyName("name")]
        public required string Name { get; set; }

        [JsonPropertyName("enabled")]
        public bool Enabled { get; set; } = true;

        [JsonPropertyName("type")]
        public required string Type { get; set; }

        /// <summary>
        /// Parameters passed to the rule engine.
        /// </summary>
        [JsonPropertyName("params")]
        public Dictionary<string, object> Params { get; set; } = new();

        /// <summary>
        /// Severity threshold (Info, Low, Medium, High, Critical).
        /// </summary>
        [JsonPropertyName("severity_threshold")]
        public string? SeverityThreshold { get; set; }

        /// <summary>
        /// Whether to skip this rule if a higher severity is found.
        /// </summary>
        [JsonPropertyName("skip_on_higher_severity")]
        public bool SkipOnHigherSeverity { get; set; } = false;
    }

    /// <summary>
    /// Output configuration - where and how to emit results.
    /// </summary>
    public sealed class OutputConfig
    {
        [JsonPropertyName("format")]
        public required string Format { get; set; } // "sarif", "json", "text"

        [JsonPropertyName("path")]
        public string Path { get; set; } = null!;

        /// <summary>
        /// Whether to include debug metadata in output.
        /// </summary>
        [JsonPropertyName("debug_metadata")]
        public bool DebugMetadata { get; set; } = false;

        /// <summary>
        /// Maximum number of findings to emit (0 = unlimited).
        /// </summary>
        [JsonPropertyName("max_findings")]
        public int MaxFindings { get; set; } = 0;
    }

    /// <summary>
    /// Rate limiting and concurrency configuration.
    /// </summary>
    public sealed class LimitsConfig
    {
        [JsonPropertyName("rate_limit")]
        public float RateLimit { get; set; } // requests per second

        [JsonPropertyName("concurrency")]
        public int Concurrency { get; set; } = 4;

        /// <summary>
        /// Retry count on transient failures.
        /// </summary>
        [JsonPropertyName("retry_count")]
        public int RetryCount { get; set; } = 3;

        /// <summary>
        /// Backoff multiplier for retries (exponential backoff).
        /// </summary>
        [JsonPropertyName("backoff_multiplier")]
        public float BackoffMultiplier { get; set; } = 2.0f;
    }

    // =====================================================================
    // CONFIG PARSER IMPLEMENTATION
    // =====================================================================

    /// <summary>
    /// Static parser for loading and validating dastlite configuration.
    /// </summary>
    public static class ConfigParser
    {
        private const string DefaultConfigPath = "dastlite.config.json";
        private const string EnvVarPrefix = "DASTLITE_";

        /// <summary>
        /// Loads configuration from a file path or default location.
        /// </summary>
        public static Config Load(string? path = null)
        {
            var fullPath = ResolveConfigPath(path);
            
            if (!File.Exists(fullPath))
            {
                Console.WriteLine($"[CONFIG] No config found at: {fullPath}");
                return CreateDefault();
            }

            try
            {
                return LoadFromStream(File.OpenRead(fullPath));
            }
            catch (JsonException ex)
            {
                throw new ConfigLoadException(
                    $"Invalid JSON in config file: {ex.Message}", 
                    ex);
            }
        }

        /// <summary>
        /// Loads configuration from an existing stream.
        /// </summary>
        public static Config LoadFromStream(Stream stream)
        {
            using var reader = new StreamReader(stream);
            return LoadFromText(reader.ReadToEnd());
        }

        /// <summary>
        /// Parses configuration text directly (useful for CLI args).
        /// </summary>
        public static Config LoadFromText(string jsonText)
        {
            try
            {
                var config = JsonSerializer.Deserialize<Config>(jsonText);
                
                if (config == null)
                    throw new ConfigLoadException("Deserialized to null - check JSON structure");

                // Apply environment variable substitution
                SubstituteEnvVars(config);

                // Validate required fields
                ValidateConfig(config);

                return config;
            }
            catch (JsonException ex)
            {
                throw new ConfigLoadException($"JSON parse error: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Creates a default configuration with sensible defaults.
        /// </summary>
        public static Config CreateDefault()
        {
            return new Config
            {
                Targets = new List<UrlTarget>
                {
                    new UrlTarget 
                    { 
                        Url = "https://example.com/api/v1",
                        Headers = { { "Accept", "application/json" } },
                        Params = { { "format", "json" } }
                    }
                },
                Authentication = new AuthConfig
                {
                    Type = "bearer_token",
                    BearerToken = "${DASTLITE_BEARER_TOKEN}"
                },
                Ruleset = new RulesetConfig
                {
                    Enabled = true,
                    DiscoverMode = false,
                    GlobalHeaders = { { "X-Client-ID", "dastlite" } }
                },
                Output = new OutputConfig
                {
                    Format = "sarif",
                    Path = "./output/dastlite.sarif",
                    MaxFindings = 1000,
                    DebugMetadata = false
                },
                Limits = new LimitsConfig
                {
                    RateLimit = 5.0f,
                    Concurrency = 4,
                    RetryCount = 3,
                    BackoffMultiplier = 2.0f
                }
            };
        }

        /// <summary>
        /// Resolves the config file path.
        /// </summary>
        private static string ResolveConfigPath(string? explicitPath)
        {
            if (!string.IsNullOrEmpty(explicitPath))
                return Path.GetFullPath(explicitPath);

            // Look in current directory, then home directory
            var candidates = new[]
            {
                DefaultConfigPath,
                Path.Combine(Environment.CurrentDirectory, DefaultConfigPath),
                Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.AppData), 
                           ".dastlite", DefaultConfigPath)
            };

            return candidates.FirstOrDefault(File.Exists) ?? 
                   Path.Combine(AppContext.BaseDirectory, DefaultConfigPath);
        }

        /// <summary>
        /// Substitutes environment variables in string values.
        /// </summary>
        private static void SubstituteEnvVars(Config config)
        {
            var prefix = config.EnvPrefix;
            
            // Helper to substitute a single string value
            Func<string, string?> subString = (s) => 
                s?.Replace(prefix, "", StringComparison.OrdinalIgnoreCase);

            // Targets
            foreach (var target in config.Targets)
            {
                target.Url = subString(target.Url);
                if (!string.IsNullOrEmpty(target.Body))
                    target.Body = subString(target.Body);
                
                foreach (var header in target.Headers)
                    header.Value = subString(header.Value);

                foreach (var param in target.Params)
                    param.Value = subString(param.Value);
            }

            // Authentication
            config.Authentication.ApiKey = subString(config.Authentication.ApiKey);
            config.Authentication.BearerToken = subString(config.Authentication.BearerToken);
            config.Authentication.OAuthClientId = subString(config.Authentication.OAuthClientId);
            config.Authentication.OAuthClientSecret = subString(config.Authentication.OAuthClientSecret);

            // Ruleset params (if any)
            foreach (var rule in config.Ruleset.Rules)
            {
                if (rule.Params != null)
                {
                    foreach (var param in rule.Params)
                        param.Value = subString(param.Value?.ToString());
                }
            }

            // Output path
            config.Output.Path = subString(config.Output.Path);
        }

        /// <summary>
        /// Validates the configuration and throws if invalid.
        /// </summary>
        private static void ValidateConfig(Config config)
        {
            var errors = new List<string>();

            // Check targets
            if (config.Targets.Count == 0)
                errors.Add("At least one target URL must be defined");

            foreach (var target in config.Targets)
            {
                if (string.IsNullOrWhiteSpace(target.Url))
                    errors.Add($"Target missing URL: {target.Method}");
                
                try 
                { 
                    new Uri(target.Url); // Will throw if invalid URI
                }
                catch (UriFormatException ex)
                {
                    errors.Add($"Invalid target URL format: {ex.Message}");
                }

                if (!string.IsNullOrEmpty(target.Body))
                {
                    var bodyLen = target.Body.Length;
                    if (bodyLen > 1024 * 1024) // 1MB limit
                        errors.Add($"Target body exceeds 1MB limit: {bodyLen} bytes");
                }
            }

            // Check authentication
            switch (config.Authentication.Type.ToLowerInvariant())
            {
                case "bearer_token":
                    if (string.IsNullOrEmpty(config.Authentication.BearerToken))
                        errors.Add("Bearer token is required for bearer_token auth type");
                    break;
                case "api_key":
                    if (string.IsNullOrEmpty(config.Authentication.ApiKey))
                        errors.Add("API key is required for api_key auth type");
                    break;
                case "oauth2":
                    if (string.IsNullOrEmpty(config.Authentication.OAuthClientId) ||
                        string.IsNullOrEmpty(config.Authentication.OAuthClientSecret))
                        errors.Add("OAuth client ID and secret are required for oauth2 auth type");
                    break;
                default:
                    errors.Add($"Unknown authentication type: {config.Authentication.Type}");
                    break;
            }

            // Check ruleset
            if (!config.Ruleset.Enabled && config.Ruleset.Rules.Count > 0)
                Console.WriteLine("[CONFIG] Ruleset disabled but contains definitions - ignoring");

            foreach (var rule in config.Ruleset.Rules)
            {
                if (string.IsNullOrWhiteSpace(rule.Id))
                    errors.Add($"Rule missing ID: {rule.Name}");
                
                if (string.IsNullOrWhiteSpace(rule.Type))
                    errors.Add($"Rule missing type: {rule.Id}");
            }

            // Check output
            if (string.IsNullOrWhiteSpace(config.Output.Format))
                config.Output.Format = "sarif"; // sensible default

            if (config.Output.MaxFindings < 0)
                config.Output.MaxFindings = 0;

            // Report errors
            if (errors.Count > 0)
            {
                Console.WriteLine($"[CONFIG] Validation found {errors.Count} issue(s):");
                foreach (var error in errors)
                    Console.WriteLine($"    - {error}");
                
                throw new ConfigLoadException("Configuration validation failed", 
                    new AggregateException(errors));
            }

            // Warn about potential issues
            if (config.Limits.RateLimit > 100.0f)
                Console.WriteLine("[CONFIG] High rate limit detected: {0} req/s may cause throttling", 
                               config.Limits.RateLimit);

            if (config.Limits.Concurrency > 32)
                Console.WriteLine("[CONFIG] High concurrency ({0}) may increase memory usage", 
                               config.Limits.Concurrency);
        }

        /// <summary>
        /// Creates a deep copy of the configuration.
        /// </summary>
        public static Config Clone(Config source)
        {
            return JsonSerializer.Deserialize<Config>(JsonSerializer.Serialize(source))!;
        }

        /// <summary>
        /// Merges two configurations, with right-hand values taking precedence.
        /// </summary>
        public static Config Merge(Config baseConfig, Config overlay)
        {
            var merged = Clone(baseConfig);

            if (overlay != null)
            {
                // Deep merge - for production use, consider a proper deep-merge utility

File: polyglot/rust/config_parser.rs
mod polyglot {
    pub mod rust {
        use std::collections::{HashMap, HashSet};
        use std::env;
        use std::fs;
        use std::path::{Path, PathBuf};
        use std::time::Duration;

        // =====================================================================
        // ERRORS & RESULT TYPES
        // =====================================================================

        #[derive(Debug)]
        pub enum ConfigError {
            Io(std::io::Error),
            Serde(serde_json::Error),
            Validation(String),
            MissingField(String),
            InvalidUrl(String),
            InvalidAuth(String),
            DuplicateProfile(String),
            EmptyConfig,
        }

        impl std::fmt::Display for ConfigError {
            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
                match self {
                    ConfigError::Io(e) => write!(f, "IO error: {}", e),
                    ConfigError::Serde(e) => write!(f, "JSON parse error: {}", e),
                    ConfigError::Validation(msg) => write!(f, "Validation failed: {}", msg),
                    ConfigError::MissingField(field) => {
                        write!(f, "Required field missing: {}", field)
                    }
                    ConfigError::InvalidUrl(url) => write!(f, "Invalid URL: {}", url),
                    ConfigError::InvalidAuth(msg) => write!(f, "Invalid auth config: {}", msg),
                    ConfigError::DuplicateProfile(name) => {
                        write!(f, "Duplicate profile name: {}", name)
                    }
                    ConfigError::EmptyConfig => write!(f, "Configuration is empty"),
                }
            }
        }

        impl std::error::Error for ConfigError {
            fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
                match self {
                    ConfigError::Io(e) => Some(e),
                    ConfigError::Serde(e) => Some(e),
                    _ => None,
                }
            }
        }

        pub type Result<T> = std::result::Result<T, ConfigError>;

        // =====================================================================
        // CONFIGURATION DATA MODELS
        // =====================================================================

        #[derive(Debug, Clone, Default, serde::Deserialize)]
        pub struct Config {
            /// Global scan options
            pub global: Option<GlobalOptions>,
            
            /// Target profiles (can be multiple for multi-target runs)
            pub profiles: Vec<Profile>,
            
            /// Curated rule set to apply
            pub ruleset: Option<RuleSetConfig>,
        }

        #[derive(Debug, Clone, Default, serde::Deserialize)]
        pub struct GlobalOptions {
            pub timeout: Option<u64>,           // Default timeout in seconds
            pub retries: Option<u16>,            // Retry count per request
            pub user_agent: Option<String>,      // Custom UA header
            pub headers: HashMap<String, String>, // Additional HTTP headers
        }

        #[derive(Debug, Clone, serde::Deserialize)]
        pub struct Profile {
            /// Unique identifier for this profile
            pub name: String,
            
            /// Target URL(s) - can be comma-separated or array
            #[serde(default, rename = "url")]
            pub urls: Vec<String>,
            
            /// Authentication configuration
            pub auth: Option<AuthConfig>,
            
            /// Scope and filtering options
            pub scope: Option<ScopeOptions>,
            
            /// Rate limiting / throttling
            pub rate_limit: Option<RateLimitOptions>,
        }

        #[derive(Debug, Clone, Default, serde::Deserialize)]
        pub struct AuthConfig {
            /// Type of authentication (basic, bearer, oauth2, cookie)
            pub kind: AuthKind,
            
            /// Basic auth credentials
            #[serde(default)]
            pub basic: Option<BasicAuth>,
            
            /// Bearer token for API auth
            #[serde(default)]
            pub bearer: Option<BearerToken>,
            
            /// OAuth2 configuration
            #[serde(default)]
            pub oauth2: Option<OAuth2Config>,
            
            /// Cookie-based auth (for session management)
            #[serde(default)]
            pub cookie: Option<SessionCookie>,
        }

        #[derive(Debug, Clone, Default, serde::Deserialize)]
        pub enum AuthKind {
            #[default]
            Basic,
            Bearer,
            OAuth2,
            Cookie,
            Header,
        }

        #[derive(Debug, Clone, Default, serde::Deserialize)]
        pub struct BasicAuth {
            pub username: String,
            pub password: String,
        }

        #[derive(Debug, Clone, Default, serde::Deserialize)]
        pub struct BearerToken {
            pub token: String,
            /// Optional scopes for OAuth2 bearer tokens
            #[serde(default)]
            pub scopes: Option<Vec<String>>,
        }

        #[derive(Debug, Clone, Default, serde::Deserialize)]
        pub struct OAuth2Config {
            pub client_id: String,
            pub client_secret: String,
            pub token_url: String,
            pub auth_url: String,
            
            /// Authorization code flow settings
            #[serde(default)]
            pub authorization_code: Option<AuthorizationCode>,
        }

        #[derive(Debug, Clone, Default, serde::Deserialize)]
        pub struct AuthorizationCode {
            pub redirect_uri: String,
            pub scopes: Vec<String>,
            pub state: String,
        }

        #[derive(Debug, Clone, Default, serde::Deserialize)]
        pub struct SessionCookie {
            /// Cookie name (e.g., "session", "auth_token")
            pub name: String,
            
            /// Optional cookie value override
            #[serde(default)]
            pub value: Option<String>,
            
            /// Path restriction for the cookie
            #[serde(default)]
            pub path: Option<String>,
        }

        #[derive(Debug, Clone, Default, serde::Deserialize)]
        pub struct ScopeOptions {
            /// Include only these paths (prefix matching)
            #[serde(default)]
            pub include: Vec<String>,
            
            /// Exclude these paths (prefix matching)
            #[serde(default)]
            pub exclude: Vec<String>,
            
            /// Depth limit for crawling
            #[serde(default)]
            pub depth: Option<u32>,
        }

        #[derive(Debug, Clone, Default, serde::Deserialize)]
        pub struct RateLimitOptions {
            /// Requests per second limit
            #[serde(default)]
            pub rps_limit: Option<f64>,
            
            /// Burst size for rate limiting
            #[serde(default)]
            pub burst_size: Option<u32>,
        }

        #[derive(Debug, Clone, Default, serde::Deserialize)]
        pub struct RuleSetConfig {
            /// List of rule IDs to enable
            #[serde(default)]
            pub enabled_rules: Vec<String>,
            
            /// Severity threshold (info, low, medium, high, critical)
            #[serde(default)]
            pub severity_threshold: Option<SeverityLevel>,
            
            /// Rule-specific overrides
            #[serde(default)]
            pub rule_overrides: HashMap<String, RuleOverride>,
        }

        #[derive(Debug, Clone, Default, serde::Deserialize)]
        pub enum SeverityLevel {
            Info,
            Low,
            #[default]
            Medium,
            High,
            Critical,
        }

        #[derive(Debug, Clone, Default, serde::Deserialize)]
        pub struct RuleOverride {
            /// Enable/disable specific rule
            #[serde(default)]
            pub enabled: Option<bool>,
            
            /// Custom severity override
            #[serde(default)]
            pub severity: Option<SeverityLevel>,
            
            /// Additional parameters for the rule
            #[serde(default)]
            pub params: HashMap<String, serde_json::Value>,
        }

        // =====================================================================
        // PARSER IMPLEMENTATION
        // =====================================================================

        impl Config {
            /// Creates a new config with defaults.
            pub fn new() -> Self {
                Self {
                    global: Some(GlobalOptions::default()),
                    profiles: Vec::new(),
                    ruleset: Some(RuleSetConfig::default()),
                }
            }

            /// Validates the configuration and returns any issues found.
            pub fn validate(&self) -> Result<Vec<String>> {
                let mut errors = Vec::new();

                // Check for empty profiles
                if self.profiles.is_empty() {
                    errors.push("At least one profile must be defined".to_string());
                }

                // Validate each profile
                let mut seen_names = HashSet::new();
                for (idx, profile) in self.profiles.iter().enumerate() {
                    if !seen_names.insert(profile.name.clone()) {
                        errors.push(format!(
                            "Profile '{}' appears multiple times",
                            profile.name
                        ));
                    }

                    // Validate URLs
                    let mut valid_urls = 0;
                    for url in &profile.urls {
                        if let Err(e) = Self::validate_url(url) {
                            errors.push(format!("Profile '{}': {}", profile.name, e));
                        } else {
                            valid_urls += 1;
                        }
                    }

                    if valid_urls == 0 && !errors.iter().any(|e| e.contains(&format!("Profile '{}'", profile.name))) {
                        errors.push(format!(
                            "Profile '{}' has no valid URLs",
                            profile.name
                        ));
                    }

                    // Validate auth configuration
                    if let Some(ref auth) = profile.auth {
                        if let Err(e) = Self::validate_auth(auth, &profile.name) {
                            errors.push(e);
                        }
                    }

                    // Validate scope options
                    if let Some(ref scope) = profile.scope {
                        if let Err(e) = Self::validate_scope(scope, &profile.name) {
                            errors.push(e);
                        }
                    }
                }

                // Check for empty config
                if self.profiles.is_empty() && self.ruleset.is_none() {
                    errors.push("No profiles or ruleset defined".to_string());
                }

                Ok(errors)
            }

            /// Validates a single URL string.
            fn validate_url(url: &str) -> Result<()> {
                if url.trim().is_empty() {
                    return Err(ConfigError::InvalidUrl("Empty URL".to_string()));
                }

                // Basic URL validation without external dependencies
                let trimmed = url.trim();
                
                // Check for scheme
                if !trimmed.starts_with("http://") && !trimmed.starts_with("https://") {
                    return Err(ConfigError::InvalidUrl(format!(
                        "Missing scheme (http:// or https://): {}",
                        trimmed
                    )));
                }

                Ok(())
            }

            /// Validates authentication configuration.
            fn validate_auth(auth: &AuthConfig, profile_name: &str) -> Result<()> {
                match &auth.kind {
                    AuthKind::Basic => {
                        if auth.basic.is_none() {
                            return Err(ConfigError::InvalidAuth(
                                format!("Profile '{}': Basic auth selected but no credentials provided", profile_name),
                            ));
                        }
                        let basic = auth.basic.as_ref().unwrap();
                        if basic.username.trim().is_empty() || basic.password.trim().is_empty() {
                            return Err(ConfigError::InvalidAuth(
                                format!("Profile '{}': Basic auth requires non-empty username and password", profile_name),
                            ));
                        }
                    }
                    AuthKind::Bearer => {
                        if auth.bearer.is_none() {
                            return Err(ConfigError::InvalidAuth(
                                format!("Profile '{}': Bearer auth selected but no token provided", profile_name),
                            ));
                        }
                        let bearer = auth.bearer.as_ref().unwrap();
                        if bearer.token.trim().is_empty() {
                            return Err(ConfigError::InvalidAuth(
                                format!("Profile '{}': Bearer token cannot be empty", profile_name),
                            ));
                        }
                    }
                    AuthKind::OAuth2 => {
                        if auth.oauth2.is_none() {
                            return Err(ConfigError::InvalidAuth(
                                format!("Profile '{}': OAuth2 selected but no configuration provided", profile_name),
                            ));
                        }
                        let oauth = auth.oauth2.as_ref().unwrap();
                        
                        // Required fields for OAuth2
                        if oauth.client_id.trim().is_empty() {
                            return Err(ConfigError::InvalidAuth(
                                format!("Profile '{}': OAuth2 client_id required", profile_name),
                            ));
                        }
                        if oauth.token_url.trim().is_empty() {
                            return Err(ConfigError::InvalidAuth(
                                format!("Profile '{}': OAuth2 token_url required", profile_name),
                            ));
                        }
                    }
                    AuthKind::Cookie => {
                        if auth.cookie.is_none() {
                            return Err(ConfigError::InvalidAuth(
                                format!("Profile '{}': Cookie auth selected but no configuration provided", profile_name),
                            ));
                        }
                        let cookie = auth.cookie.as_ref().unwrap();
                        if cookie.name.trim().is_empty() {
                            return Err(ConfigError::InvalidAuth(
                                format!("Profile '{}': Cookie name required", profile_name),
                            ));
                        }
                    }
                    AuthKind::Header => {
                        // Header-based auth is minimal - just warn if no headers specified
                        if auth.basic.is_none() && auth.bearer.is_none() {
                            return Err(ConfigError::InvalidAuth(
                                format!("Profile '{}': Header auth selected but no credentials provided", profile_name),
                            ));
                        }
                    }
                }

                Ok(())
            }

            /// Validates scope options.
            fn validate_scope(scope: &ScopeOptions, profile_name: &str) -> Result<()> {
                // Validate include/exclude patterns
                for pattern in &scope.include {
                    if pattern.trim().is_empty() {
                        return Err(ConfigError::Validation(format!(
                            "Profile '{}': Include pattern cannot be empty",
                            profile_name
                        )));
                    }
                }

                for pattern in &scope.exclude {
                    if pattern.trim().is_empty() {
                        return Err(ConfigError::Validation(format!(
                            "Profile '{}': Exclude pattern cannot be empty",
                            profile_name
                        )));
                    }
                }

                // Validate depth
                let depth = scope.depth.unwrap_or(10);
                if depth > 50 {
                    return Err(ConfigError::Validation(format!(
                        "Profile '{}': Depth limit of {} is unusually high",
                        profile_name, depth
                    )));
                }

                Ok(())
            }

            /// Applies default values for missing optional fields.
            pub fn apply_defaults(&mut self) {
                // Apply global defaults
                if let Some(ref mut global) = self.global {
                    if global.timeout.is_none() {
                        global.timeout = Some(30); // 30 second default timeout
                    }
                    if global.retries.is_none() {
                        global.retries = Some(3);
                    }
                }

                // Apply profile defaults
                for profile in &mut self.profiles {
                    // Default scope options
                    if profile.scope.is_none() {
                        profile.scope = Some(ScopeOptions::default());
                    }

                    // Default rate limiting (10 req/s burst of 50)
                    if profile.rate_limit.is_none() {
                        profile.rate_limit = Some(RateLimitOptions {
                            rps_limit: Some(10.0),
                            burst_size: Some(50),
                        });
                    }

                    // Default auth (no auth by default, which is fine for public APIs)
                    if profile.auth.is_none() {
                        profile.auth = Some(AuthConfig::default());
                    }
                }

                // Apply ruleset defaults
                if let Some(ref mut ruleset) = self.ruleset {
                    // Default severity threshold
                    if ruleset.severity_threshold.is_none() {
                        ruleset.severity_threshold = Some(SeverityLevel::Medium);
                    }
                }
            }

            /// Merges environment variables into the config.
            pub fn merge_env(&mut self) -> Result<()> {
                // Environment variable interpolation for URLs and auth
                let env_prefix = "DASTLITE_";
                
                // Merge global options from env
                if let Some(ref mut global) = self.global {
                    if let Ok(val) = env::var(format!("{}_TIMEOUT", env_prefix)) {
                        if !val.trim().is_empty() {
                            global.timeout = Some(val.parse::<u64>().unwrap_or(30));
                        }
                    }

                    if let Ok(val) = env::var(format!("{}_RETRIES", env_prefix)) {
                        if !val.trim().is_empty() {
                            global.retries = Some(val.parse::<u16>().unwrap_or(3));
                        }
                    }
                }

                // Merge profile-specific env vars
                for (idx, profile) in self.profiles.iter_mut().enumerate()